### PR TITLE
Features/714 trace

### PR DESCRIPTION
**Pull request by lenablind**
## Description

Implementation of `ht.trace`, which computes the sum along the diagonal, analogous to [np.trace](https://numpy.org/doc/stable/reference/generated/numpy.trace.html).

Used torch functions: 
- [torch.trace](https://pytorch.org/docs/stable/generated/torch.trace.html) 
- [torch.diagonal](https://pytorch.org/docs/stable/generated/torch.diagonal.html)
- [torch.sum](https://pytorch.org/docs/stable/generated/torch.sum.html)

Issue/s resolved: #714

### Strategy (Algorithm)
**CASE 2D Input**
*Returns*: Scalar

- If offset would result into an empty array, directly create a 0d-tensor with value 0
- Otherwise, adapt the offset to the distribution,
- Defer the resulting sub-arrays, upholding the affected diagonal entries then
- Calculate the trace of each sub-array and 
- Sum up all partial results by means of an `MPI_Allreduce` operation to obtain the final result

**CASE > 2D Input**
*Returns*: DNDArray
(Hint: as `torch.trace` is only implemented for 2-dimensional tensors, I decided to do a workaround by first extracting the diagonal entries via `torch.diagonal` followed by a summation along the last axis of the resulting array)

I differentiated between two cases: the split axis being or not being within the "trace axes" (`axis1`, `axis2`), as this leads to the required diagonal entries of the last axis being distributed on varying processes or not.

*CASE: split axis NOT in (`axis1`, `axis2`)*
- directly extract the diagonal of the data chunk
- compute the sum along the last axis via `torch.sum` as explained above
- set the partial results back together along the correct axis ("`gather_axis`")

*CASE: split axis IN (`axis1`, `axis2`)*
- adapt the offset to the distribution*
- (*: if it would not result into an empty array, otherwise `torch.diagonal` provides such even with unmodified `offset`)
- extract the diagonal entries
- adapt the shape of this array to the shape of result
- sum up all elements by means of `MPI_Allgather`

If `out` is not `None`, provide the correct split configuration to store the result in the given variable.

## Changes proposed:
- Implementation of `ht.trace`
- Implementation of corresponding test

## Type of change
- New feature (non-breaking change which adds functionality)

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by coquelin77** [comment]
rerun tests

File: heat/core/linalg/basics.py [hunk 1]
```diff
@@ -1075,6 +1077,317 @@ def projection(a, b):
...
+            # Store result in out if provided
+            if out is not None:
+                if out.split is not None:
+                    warnings.warn(
+                        f"Split axis of `out` will be changed from {out.split} to None to "
+                        f"guarantee correct results."
+                    )
+                    # Allgather of out - necessary as whole result is on all processes after Allreduce
+                    out.resplit_(None)
+                sanitation.sanitize_out(out, tuple(res_shape), out.split, a.device)
```
Comment (coquelin77) [inline heat/core/linalg/basics.py:1326]: sanitation typically at the top

File: heat/core/linalg/basics.py [hunk 2]
```diff
@@ -1075,6 +1077,317 @@ def projection(a, b):
...
+            sum_along_diagonals_t, split=a.split, dtype=dtype, device=a.device, comm=a.comm
+        )
+
+        # convert resulting 0-d DNDarray to scalar
+        return sum_along_diagonals.item()
+
+    # -------------------------------
+    # CASE > 2D => DNDArray
+    # -------------------------------
+    else:
```
Comment (coquelin77) [inline heat/core/linalg/basics.py:1268]: since the if block above ends with a return and the rest of the function is in an `else` block, this `else` can be removed and the code be one tab less

File: heat/core/linalg/basics.py [hunk 3]
```diff
@@ -1075,6 +1077,317 @@ def projection(a, b):
...
+            # check if split axis is in range of result
+            if a.split is not None and a.split >= sum_along_diagonals_t.ndim:
+                gather_axis = sum_along_diagonals_t.ndim - 1
+            else:
+                gather_axis = a.split
+
+            # convert torch result back to DNDarray
+            sum_along_diagonals = factories.array(
+                sum_along_diagonals_t, dtype=dtype, split=gather_axis, comm=a.comm, device=a.device
+            )
```
Comment (coquelin77) [inline heat/core/linalg/basics.py:1369]: to clarify, you will return a split object even if the input is `split=None`?

File: CHANGELOG.md [hunk 4]
```diff
@@ -27,6 +27,9 @@
 - [#679](https://github.com/helmholtz-analytics/heat/pull/679) New feature: ``histc()`` and ``histogram()``
 ### Linear Algebra
 - [#658](https://github.com/helmholtz-analytics/heat/pull/658) Bugfix: `matmul` on GPU will cast away from `int`s to `float`s for the operation and cast back upon its completion. This may result in numerical inaccuracies for very large `int64` DNDarrays
+- [#718](https://github.com/helmholtz-analytics/heat/pull/718) New feature: `trace()`
+### Logical
+- [#711](https://github.com/helmholtz-analytics/heat/pull/711) `isfinite()`, `isinf()`, `isnan()`
```
Comment (coquelin77) [inline CHANGELOG.md:32]: this became displaced on a merge. see the master changelog for its proper position

**Review by coquelin77 — CHANGES_REQUESTED** [review]

File: heat/core/dndarray.py [hunk 5]
```diff
@@ -3714,6 +3714,77 @@ def tolist(self, keepsplit=False) -> List:
 
         return self.__array.tolist()
 
+    def trace(self, offset=0, axis1=0, axis2=1, dtype=None, out=None):
```
[the remark this answers is not in the document]
Comment (lenablind) [reply heat/core/dndarray.py:3717]: @coquelin77 Thank you for the review! I'll look into it and adapt the changes as soon as 504 is introduced

[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/core/dndarray.py:3717]: it was merged yesterday afternoon, there are many examples of how to make the docstrings. or check out the files in `doc/source/installation.rst`

[the remark this answers is not in the document]
Comment (lenablind) [reply heat/core/dndarray.py:3717]: Alright, awesome, I'll check them out later. Thanks for the hint!

File: heat/core/linalg/basics.py [hunk 6]
```diff
@@ -1075,6 +1077,317 @@ def projection(a, b):
...
+                sum_along_diagonals_t = sum_along_diagonals_t.type(a_sub.dtype.torch_type())
+
+            # empty array => result = 0
+            else:
+                sum_along_diagonals_t = torch.tensor(
+                    0, dtype=a_sub.dtype.torch_type(), device=a_sub.device.torch_device
+                )
+
+        if out is not None:
+            raise ValueError("`out` not applicable (/must be None) if result is a scalar/ input 2D")
```
[the remark this answers is not in the document]
Comment (lenablind) [reply heat/core/linalg/basics.py]: Actually it wasn't, as this is a separate error case: Someone would like to store the results in out, which is not possible given a 2-dimensional input resulting into a scalar output (numpy throws an error here, so did I). I put the sanitation at the top of the function though

File: heat/core/linalg/basics.py [hunk 7]
```diff
@@ -1075,6 +1077,317 @@ def projection(a, b):
...
+        if a.is_distributed():
+            a.comm.Allreduce(MPI.IN_PLACE, sum_along_diagonals_t, MPI.SUM)
+
+        # cast to DNDarray of correct dtype
+        sum_along_diagonals = factories.array(
+            sum_along_diagonals_t, split=a.split, dtype=dtype, device=a.device, comm=a.comm
+        )
+
+        # convert resulting 0-d DNDarray to scalar
+        return sum_along_diagonals.item()
```
[the remark this answers is not in the document]
Comment (lenablind) [reply heat/core/linalg/basics.py]: You are right, this conversion is an unnecessary step -  I removed it

File: heat/core/linalg/basics.py (hunk 3 above)
Comment (lenablind) [reply heat/core/linalg/basics.py]: No, this won't happen. If `a.split=None`, the first part of the (`and`) if condition is already false and the else case is always executed. Thus the result will also be not distributed.

**Comment by coquelin77** [comment]
rerun tests

**Comment by coquelin77** [comment]
rerun tests

**Review by coquelin77 — APPROVED** [review]
